### DISCUSSION TITLE
`jupyverse` as a backend

### DISCUSSION THREAD

**Discussion by jtpio** [jupyterlite/jupyterlite#372] [Ideas]
At the moment JupyterLite implements its own Jupyter Server in TypeScript: https://github.com/jupyterlite/jupyterlite/tree/main/packages/server

This is quite convenient and has been working quite ok for now. It's quite easy to integrate with browser APIs for example to store content to the local storage.

However since `fastapi` and `fps` can be installed with `micropip` it would be tempting to reuse them for the JupyterLite server:

![image](https://user-images.githubusercontent.com/591645/136048382-7dc0f6cd-a2d9-43e8-bd99-b30664618622.png)

## Pros

- Reuse the same code as `jupyverse`: 1 Jupyter Server implementation that can run on servers and in the browser
- Easier to reuse server plugins 

## Cons

- Heavy start. It would be a downgraded experience if we had to wait for Pyodide to finish loading before being able to use JupyterLite. Although `ServiceWorker`s might help here (#137)
- The JS -> Python -> JS round trip might be awkward if we want to start JavaScript kernels
- Interface with the browser API in Python might also feel clunky?

**Comment by bollwyvl** [comment]
I see a hard, first-party pyodide dependency for starting _any_  other UI would be pretty serious net loss, even if we did have archival-grade deployable python dependencies from #310 on the way to #137. For non-python languages (e.g. plain-old-js, #96) this would incur an order of magnitude more requests and total download size. Getting to re-use the battle-tested upstream `@jupyterlab` types is probably worth the price of admission for typescript, etc.

However, as we consider making more of core components pluggable, it would certainly be an option for a third party to implement and test against such a contraption.

As to re-using server components: yeah, i dunno. There's _just_ enough stuff that isn't going to work the right way, and things like databases, sensible network comms, etc. are going to be _right out_ for the time being.

**Comment by jtpio** [comment]
Commenting as not planning to do anything about this for now.

For reference there is also a discussion in https://github.com/jupyterlite/jupyterlite/issues/779.
